### PR TITLE
feat: make StreamRange<T> public

### PR DESCRIPTION
**Pull request by devjgm**
This class is moving from `google/cloud/internal/` to `google/cloud/`, and also in a non-internal namespace. This is now a user-facing class that we want users to be able to use and read the documentation for. We expect to have user-facing functions _directly_ returning objects of type `google::cloud::StreamRange<Foo>`, so it needs to be public.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5811)

### CODE REVIEW

File: google/cloud/stream_range.h [hunk 1]
```diff
@@ -85,13 +91,13 @@ class StreamRange {
   template <typename U>
   static constexpr bool IsMoveNoexcept() {
     return noexcept(StatusOr<U>(std::declval<U>()))&& noexcept(
-        StreamReader<U>(std::declval<StreamReader<U>>()));
+        internal::StreamReader<U>(std::declval<internal::StreamReader<U>>()));
   }
 
  public:
   /// An input iterator for a `StreamRange<T>`
```
Comment (coryan) [inline google/cloud/stream_range.h:98]: nit: maybe remind users that this is not part of the public API and they should be using the `iterator` alias below.

Comment (devjgm) [reply google/cloud/stream_range.h]: Good idea. Done.

File: google/cloud/stream_range.h [hunk 2]
```diff
@@ -186,22 +196,46 @@ class StreamRange {
...
+   * end of stream, or a non-OK `Status` to indicate an error, or a `T`. The
+   * `internal::StreamReader` will not be invoked again after it returns a
+   * `Status`.
+   *
+   * @par Example: Printing integers from 1-10.
+   *
+   * @code
+   * int counter = 0;
+   * auto reader = [&counter]() -> internal::StreamReader<int>::result_type {
+   *   if (counter++ < 10) return counter;
```
Comment (devbww) [inline google/cloud/stream_range.h:216]: Nit for moved example: `if (++counter <= 10) return counter;` is a little nicer.

**Review by coryan — APPROVED** [review]
:lgtm:

> Reviewed 5 of 6 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5811#-:-MSjNpAF463gcDs18nbb:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @devjgm)

File: google/cloud/stream_range.h (hunk 2 above)
Comment (devjgm) [reply google/cloud/stream_range.h]: Changed here and above.

**Pull request merged**